Move check for future timestamps from HLC.Update to Store.ExecuteCmd

bdarnell opened the pull request:
HLC.Update is called post-raft, which means it must not vary based on
the node's current clock.

See #1242.

bdarnell committed:
Move check for future timestamps from HLC.Update to Store.ExecuteCmd.

HLC.Update is called post-raft, which means it must not vary based on
the node's current clock.

See #1242.

bdarnell committed:
Remove error return from hlc.Clock.Update.

tamird commented:
Is it possible to add a test for the high-level symptom this is fixing?

tschottdorf commented on kv/txn_test.go:
```diff
@@ -201,9 +199,7 @@ func verifyUncertainty(concurrency int, maxOffset time.Duration, t *testing.T) {
 			// hasn't changed (i=0). The logical component will change
 			// internally in a way we can't track, but we want to be just
 			// ahead.if
```
that comment is funky.

bdarnell replied:
Yeah. I can remove the "if" at the end but I'm not sure what the comment as a whole is trying to say.

tschottdorf replied:
We want to avoid having the internal clock tick ahead of our timestamp here, so we blindly add 999 ticks hoping that's enough. Should probably take another look at it; it doesn't seem very sane.

tschottdorf commented:
LGTM, but a small test would be nice.

spencerkimball commented:
Nice. LGTM

vivekmenezes commented on util/hlc/hlc.go:
```diff
@@ -193,12 +193,9 @@ func (c *Clock) PhysicalTime() time.Time {
-	defer func() {
-		result = c.timestamp()
-	}()
 	physicalClock := c.physicalClock()
```
might as well remove the defer result computation from Now() as well

bdarnell replied:
Huh, that's just Now() being very weird. I'll make it a normal return.

bdarnell commented:
I'm working on a test and finding bugs. It looks like we run into trouble when using three stores with separate manual clocks, since the leader's clock will not advance and it will retry transactions at the same timestamp. Our current tests either use the physical clock which will advance on its own, or a single manual clock shared by all tests (in which case when the followers update their clock to leader + 1 the leader's clock changes too). I think we are missing a call to clock.Update somewhere to ensure we never reuse timestamps but I'm not sure where it belongs.

bdarnell committed:
Add tests for timestamp update behavior in Store.

The tests verify that time jumps greater than MaxOffset are allowed if
they make it past the raft phase, but rejected if caught earlier.

bdarnell commented:
After a lengthy digression to fix some issues in the way we handle timestamps, I've added tests in 4889c66. I've included commits from #1326 here; I'll rebase them away once that PR is merged.

bdarnell committed:
PR feedback
